Neos.Media/Classes/Domain/Service/AssetVariantGenerator.php

```diff
@@ -0,0 +1,118 @@
+<?php
+declare(strict_types=1);
+
+namespace Neos\Media\Domain\Service;
+
+/*
+ * This file is part of the Neos.Media package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Configuration\Exception\InvalidConfigurationException;
+use Neos\Flow\ResourceManagement\Exception;
+use Neos\Media\Domain\Model\Adjustment\ImageAdjustmentInterface;
+use Neos\Media\Domain\Model\AssetInterface;
+use Neos\Media\Domain\Model\AssetVariantInterface;
+use Neos\Media\Domain\Model\Image;
+use Neos\Media\Domain\Model\ImageVariant;
+use Neos\Media\Domain\ValueObject\Configuration;
+use Neos\Media\Exception\AssetVariantGeneratorException;
+use Neos\Media\Exception\ImageFileException;
+use Neos\Utility\ObjectAccess;
+
+/**
+ * @Flow\Scope("singleton")
+ */
+class AssetVariantGenerator
+{
+    /**
+     * @Flow\Inject
+     * @var AssetService
+     */
+    protected $assetService;
+
+    /**
+     * @param AssetInterface $asset
+     * @return ImageVariant[] The created variants (if any), with the preset identifier as array key
+     * @throws AssetVariantGeneratorException
+     * @throws Exception
+     * @throws ImageFileException
+     * @throws InvalidConfigurationException
+     */
+    public function createVariants(AssetInterface $asset): array
+    {
+        // Currently only Image Variants are supported. Other asset classes can be supported, as soon as there is a common
+        // interface for creating and adding variants.
+        //
+        if (!$asset instanceof Image) {
+            return [];
+        }
+
+        $createdVariants = [];
+        foreach ($this->assetService->getVariantPresets() as $presetIdentifier => $preset) {
+            if ($preset->matchesMediaType($asset->getMediaType())) {
+                foreach ($preset->variants() as $variantConfiguration) {
+                    $createdVariants[$presetIdentifier] = $this->createVariant($asset, $presetIdentifier, $variantConfiguration);
+                    $asset->addVariant($createdVariants[$presetIdentifier]);
```

Comment:
If I am not mistaken to be unique in the `$createdVariants` array the key must be comprised of `$presetIdentifier` and `$variantConfiguration->identifier()` ?

Reply:
Good catch! I didn't realize this because the return value is actually only used in the unit test.